### DISCUSSION TITLE
Code folding in PDF output?

### DISCUSSION THREAD

**Discussion by domilot** [quarto-dev/quarto-cli#4121] [Feature Requests]
I just tried Quarto with a Jupyter notebook. The code folding feature seems to work exclusively for HTML output. Unfortunately, this feature is missing from where it would seem really necessary, i.e. fixed-width documents like PDF. And this in a project promising "elegantly formatted output", "high-quality articles", and "beautiful publications" (all taken from Quarto's homepage).

**Comment by cscheid** [comment]
The PDF format standard itself is static, and doesn't support interactive features. You can still control whether code is hidden with `echo: false`, `output: false`, etc.

**Reply by domilot** [reply]
Sure, PDF is static. This is precisely why it would be nice to have code automatically wrapped prior to the PDF-file generation. Otherwise this is what you get:
[too_long_line.pdf](https://github.com/quarto-dev/quarto-cli/files/10519669/too_long_line.pdf)

**Reply by cscheid** [reply]
Ok, I see what you mean. This confused me:

<img width="1012" alt="image" src="https://user-images.githubusercontent.com/285675/215114590-a2728832-dbdf-44c9-bc66-d33c4f057b3e.png">

> The code folding feature

We'd like to do better with wrapping as well. The problem is that the type formatting that needs to be done is dependent on the final PDF format, which we don't know about at time of rendering. So it's actually a hard problem to solve in general. You see the same problem in HTML formats as well. I believe you'll find that this isn't a problem unique to quarto.

**Reply by domilot** [reply]
I know this is a hard problem to solve. I got the same ugly results in the past while exporting my RST (restructured text) files to PDF via Sphinx. I keep my fingers crossed for your team because otherwise Quarto seems to do a great job.

**Reply by dragonstyle** [reply]
I think automatically wrapping code in a satisfactory way is a pretty tough problem. Often the author having control of the wrapping behavior is preferable as it allows the placement of breaks in the most sensible locations (whereas automatic wrapping is unlikely to achieve this).

**Comment by baptiste** [comment]
There are [a few pointers on the knitr side of things](https://github.com/rstudio/rmarkdown/issues/646); this one in particular seems to work well

````
---
format: 
  pdf:
    keep-tex: true
    include-in-header: 
       text: |
         \usepackage{fvextra}
         \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
         \DefineVerbatimEnvironment{OutputCode}{Verbatim}{breaklines,commandchars=\\\{\}}
---

This is Markdown text. The line is extremely long. Luckily, it gets wrapped automatically, which is very, very nice.
   

```{r}
#| class-output: OutputCode
grid::unit(lattice::trellis.par.get()$axis.components$bottom$tck, lattice::lattice.options()$layout.widths$right.padding$units)

print("This is a very long line of code that I would gladly see folded. Unfortunately it does not appear my screen expands as needed, which is a shame as I could not preview Hasselblad XPan scans, or look at xkcd 1732 sideways.")
```

````

![Screenshot 2023-01-28 at 09 31 56](https://user-images.githubusercontent.com/18970/215191436-d07e86ad-e248-430b-8a11-b7ce2261ace6.png)


However I do not understand the behaviour of `class-output` (and `attr-output` alongside, which turns things into a "html-div" but not in a way understood by LaTeX); with just `class-output` it switches on the Highlighting + Shaded environment no matter the name, but I'm not sure how to get a plain Verbatim one for the output.

**Comment by CLRafaelR** [comment]
Although [baptiste's solution](https://github.com/quarto-dev/quarto-cli/discussions/4121#discussioncomment-4800510) works very fine, you can avoid setting `class-output` option for each code block if you redefine `verbatim` environment using `\RecustomVerbatimEnvironment{verbatim}{Verbatim}{...}` in `include-before-body` key of YAML. See also [a relevant thread](https://tex.stackexchange.com/q/468220/169454) in tex.stackexchange.

````
---
format: 
  pdf:
    keep-tex: true
    include-in-header: 
       text: |
         \usepackage{fvextra}
         \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
include-before-body:
  text: |
    \RecustomVerbatimEnvironment{verbatim}{Verbatim}{
      showspaces = false,
      showtabs = false,
      breaksymbolleft={},
      breaklines
      % Note: setting commandchars=\\\{\} here will cause an error 
    }
---

This is Markdown text. The line is extremely long. Luckily, it gets wrapped automatically, which is very, very nice.
   

```{r}
grid::unit(lattice::trellis.par.get()$axis.components$bottom$tck, lattice::lattice.options()$layout.widths$right.padding$units)

print("This is a very long line of code that I would gladly see folded. Unfortunately it does not appear my screen expands as needed, which is a shame as I could not preview Hasselblad XPan scans, or look at xkcd 1732 sideways.")
```
````

![image](https://user-images.githubusercontent.com/42314553/236612696-b4be012c-898a-4707-b18c-1db59bff7cfa.png)

**Reply by Pierrrick** [reply]
How would one deal with this kind of output not fitting into the page ?

![image](https://user-images.githubusercontent.com/48405156/236614756-2d6b84a3-cba2-4710-97a3-3c8477db448a.png)

**Reply by mcanouil** [reply]
@Pierrrick Please try to post question relevant to the thread after trying the solution.
Also in your case, since you are asking about code output wrapping, see #3693 for example.

**Reply by CLRafaelR** [reply]
@Pierrrick 

My solution works for folding long message output:

![image](https://user-images.githubusercontent.com/42314553/236623253-7f49c756-7290-460f-ae13-02384796fb23.png)

Did you really try the solutions posted here? If so, you ought to show us what you did and what environment you are working with (e.g. run `sessionInfo()` and retrieve R version, versions of relevant packages, and your OS), then we can help you.

````
---
format: 
  pdf:
    keep-tex: true
    include-in-header: 
       text: |
         \usepackage{fvextra}
         \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
include-before-body:
  text: |
    \RecustomVerbatimEnvironment{verbatim}{Verbatim}{
      showspaces = false,
      showtabs = false,
      breaksymbolleft={},
      breaklines
    }
---

This is Markdown text. The line is extremely long. Luckily, it gets wrapped automatically, which is very, very nice.
   

```{r}
grid::unit(lattice::trellis.par.get()$axis.components$bottom$tck, lattice::lattice.options()$layout.widths$right.padding$units)

print("This is a very long line of code that I would gladly see folded. Unfortunately it does not appear my screen expands as needed, which is a shame as I could not preview Hasselblad XPan scans, or look at xkcd 1732 sideways.")
```


```{r}
library(tidyverse)
```
````

**Comment by skwde** [comment]
This did not work for me.

I had to add use

```yml
    include-in-header: 
      text: |
        \usepackage{fvextra}
        \DefineVerbatimEnvironment{Highlighting}{Verbatim}{
          commandchars=\\\{\},
          breaklines, breaknonspaceingroup, breakanywhere
        }
```

instead.

See also the dedicated pandoc latex section in the `fvextra` docs: https://mirror.init7.net/ctan/macros/latex/contrib/fvextra/fvextra.pdf

**Reply by leowill01** [reply]
this worked for me after trying multiple different things. thanks!

**Comment by s2t2** [comment]
How to do this for Python?

**Reply by mcanouil** [reply]
What do you mean for Python?
Nothing discussed here depends on the language.

**Comment by averms** [comment]
Typst does this automatically, you can use `pdf-engine: typst`

**Reply by mcanouil** [reply]
@averms Quarto supports Typst as a format. Use `format: typst`.

**Comment by cdriveraus** [comment]
I had a lot of headaches with this and couldn't find an adequate solution, so posting the one that worked for me in case others end up here:

````
---
title: "wrap text"
format:
  pdf:
    keep-tex: true
    include-in-header: 
      - text: |
          \usepackage{fvextra}
          \fvset{breaklines=true, commandchars=\\\{\}, breaksymbolleft={}, breakindentnchars=8}

---

```{r}
#| label: setup
verbwrap <- function(x, options) {
  paste("\\begin{Verbatim}\n",
        x, "\\end{Verbatim}\n", sep = "")
}
knitr::knit_hooks$set(output = verbwrap)
```

```{r}
print("I just want this text to wrap to a new line I just want this text to wrap to a new line I just want this text to wrap to a new line ")
```

````

**Reply by mcanouil** [reply]
Thanks for sharing!
I edited the formatting of the post.

Note that in Quarto, in code cells, there should be only the language inside the brackets. Everything else should be in YAML with the comment+pipe prefix.
